src/reporter.cc

```diff
@@ -0,0 +1,183 @@
+  request << boost::network::header("Content-Length",
+                                    std::to_string(request_body.size()));
+  request << boost::network::header("Content-Type", "application/json");
+  request << boost::network::header("Authorization", auth_header);
+  request << boost::network::body(request_body);
+  http::client::response response = client.post(request);
+  if (status(response) >= 300) {
+    throw boost::system::system_error(
+        boost::system::errc::make_error_code(boost::system::errc::not_connected),
+        format::Substitute("Server responded with '{{message}}' ({{code}})",
+                           {{"message", status_message(response)},
+                            {"code", format::str(status(response))}}));
+  }
+  if (verbose_logging) {
+    LOG(INFO) << "Server responded with " << body(response);
+  }
+  // TODO: process response.
+}
+
+}
+
+void MetadataReporter::SendMetadata(
+    std::map<MonitoredResource, MetadataStore::Metadata>&& metadata)
+    throw (boost::system::system_error) {
+  if (metadata.empty()) {
+    if (config_.VerboseLogging()) {
+      LOG(INFO) << "No data to send";
+    }
+    return;
+  }
+
+  if (config_.VerboseLogging()) {
+    LOG(INFO) << "Sending request to the server";
+  }
+  const std::string project_id = environment_.NumericProjectId();
+  // The endpoint template is expected to be of the form
+  // "https://stackdriver.googleapis.com/.../projects/{{project_id}}/...".
+  const std::string endpoint =
+      format::Substitute(config_.MetadataIngestionEndpointFormat(),
+                         {{"project_id", project_id}});
+  const std::string auth_header = auth_.GetAuthHeaderValue();
+  const std::string user_agent = config_.MetadataReporterUserAgent();
+
+  const json::value empty_request = json::object({
+    {"entries", json::array({})},
+  });
+  const int empty_size = empty_request->ToString().size();
+
+  const int limit_bytes = config_.MetadataIngestionRequestSizeLimitBytes();
+  int total_size = empty_size;
+
+  std::vector<json::value> entries;
+  for (auto& entry : metadata) {
+    const MonitoredResource& resource = entry.first;
+    MetadataStore::Metadata& metadata = entry.second;
+    if (metadata.ignore) {
+      continue;
+    }
+    json::value metadata_entry =
+        json::object({  // MonitoredResourceMetadata
+          {"resource", resource.ToJSON()},
+          {"rawContentVersion", json::string(metadata.version)},
+          {"rawContent", std::move(metadata.metadata)},
+          {"state", json::string(metadata.is_deleted ? "DELETED" : "ACTIVE")},
+          {"createTime", json::string(time::rfc3339::ToString(metadata.created_at))},
+          {"collectTime", json::string(time::rfc3339::ToString(metadata.collected_at))},
+        });
+    // TODO: This is probably all kinds of inefficient...
+    const int size = metadata_entry->ToString().size();
+    if (empty_size + size > limit_bytes) {
```

Comment:
Just curious, do we expect this to happen often? Do you know what likely scenarios would exist where this would be exceeded frequently? I'll setup log metrics for this line, and try to reproduce the event, so that we can get a handle on frequency once I hear from you.

Reply:
This is just defensive programming to catch individual metadata entries that would take the whole request over the size limit. We haven't seen any so far, but given that we include the raw output of an unrelated API, this isn't too outrageous of a scenario in my mind.